src/utils/sllippage_model_fill_adjustor.ts

```diff
@@ -0,0 +1,51 @@
+import { BigNumber, Fill, FillAdjustor, FillData, adjustOutput } from '@0x/asset-swapper';
+import { MarketOperation } from '@0x/types';
+import { SlippageModelManager } from './slippage_model_manager';
+
+export class SlippageModelFillAdjustor implements FillAdjustor {
+    constructor(
+        private readonly slippageModelManager: SlippageModelManager,
+        private readonly sellToken: string,
+        private readonly buyToken: string,
+        private readonly maxSlippageRate: number,
+    ) {}
+
+    public adjustFills(side: MarketOperation, fills: Fill<FillData>[], amount: BigNumber): Fill<FillData>[] {
+        return fills.map((f) => {
+            // Mostly negative, as in the trade experiences a worst price
+            // e.g -0.02938
+            const expectedSlippage = this.slippageModelManager.calculateExpectedSlippage(
+                this.buyToken,
+                this.sellToken,
+                side === MarketOperation.Sell ? f.output : f.input,
+                side === MarketOperation.Sell ? f.input : f.output,
+                [{ name: f.source, proportion: new BigNumber(1) }],
+                this.maxSlippageRate,
+            );
+
+            if (expectedSlippage === null) {
+                return f;
+            }
+
+            const fill = { ...f };
+            // 1000 * (1 + -0.02938) = 970
+            const slippageAdjustedOutput = f.output
```

Comment:
Shall the `slippageAdjustedOutput` depend on `side`? Seems the adjust will always decrease `output` amount as `expectedSlippage` is mostly negative. But for sell the output is the higher the better, while for buy the output is the lower the better for takers.